model/task/task.go

```diff
@@ -2486,82 +2487,90 @@ func (t *Task) Insert() error {
+			bson.M{
+				"$unset": bson.M{
+					AbortedKey:              "",
+					AbortInfoKey:            "",
+					OverrideDependenciesKey: "",
+				},
+				"$set": bson.M{
+					ExecutionKey:             t.Execution + 1,
+					LatestParentExecutionKey: t.Execution + 1, // Will this matter for regular tasks?
+				},
+			})
+		if err != nil {
+			return errors.Wrap(err, "updating task")
 		}
+		t.Aborted = false
+		return nil
 	}
+}
 
-	archiveTask := t.makeArchivedTask()
-	err := db.Insert(OldCollection, archiveTask)
-	if err != nil {
-		grip.Error(message.WrapError(err, message.Fields{
-			"archive_task_id": archiveTask.Id,
-			"old_task_id":     archiveTask.OldTaskId,
-			"execution":       t.Execution,
-			"display_only":    t.DisplayOnly,
-		}))
-		return errors.Wrap(err, "inserting archived task into old tasks")
-	}
-	err = UpdateOne(
-		bson.M{IdKey: t.Id},
-		bson.M{
-			"$unset": bson.M{
-				AbortedKey:              "",
-				AbortInfoKey:            "",
-				OverrideDependenciesKey: "",
-			},
-			"$inc": bson.M{ExecutionKey: 1},
-		})
-	if err != nil {
-		return errors.Wrap(err, "updating task")
-	}
-	t.Aborted = false
-	return nil
+func shouldRestartTask(resetFailedWhenFinished bool, resetWhenFinished bool, status string) bool {
+	return resetWhenFinished || !resetFailedWhenFinished || evergreen.IsFailedTaskStatus(status)
 }
 
+// ArchiveMany accepts tasks and display tasks (no execution tasks), bundles them all up, archives and updates all
+// that need to be (accounting for ResetFailedWhenFinished) in two queries
 func ArchiveMany(tasks []Task) error {
-	if len(tasks) == 0 {
-		return nil
-	}
-	existingTasksMap := map[string]bool{}
-	for _, t := range tasks {
-		existingTasksMap[t.Id] = true
-	}
-	additionalTasks := []string{}
+	bundledTasks := []string{}        // Contains all tasks and display tasks
+	execTasks := []string{}           // Contains all exec tasks
+	toUpdateExecTaskIds := []string{} // Contains all exec tasks that should update and have new execution
+	toArchive := []interface{}{}      // Contains all archived tasks that are being updated / have new execution
+
 	for _, t := range tasks {
-		// For any display tasks here, make sure that we also archive all their execution tasks.
-		// Consider each execution tasks individually, so we don't query a task we already passed in.
-		for _, et := range t.ExecutionTasks {
-			if !existingTasksMap[et] {
-				additionalTasks = append(additionalTasks, et)
-				existingTasksMap[et] = true
+		bundledTasks = append(bundledTasks, t.Id)
+		toArchive = append(toArchive, t.makeArchivedTask())
+		if t.DisplayOnly && len(t.ExecutionTasks) > 0 {
+			eTasks, err := FindAll(db.Query(ByIds(t.ExecutionTasks)))
+			if err != nil {
+				return errors.Wrapf(err, "finding execution tasks for display task '%s'.", t.Id)
+			}
+			for _, et := range eTasks {
+				execTasks = append(execTasks, et.Id)
+				if shouldRestartTask(t.ResetFailedWhenFinished, t.ResetWhenFinished, et.Status) {
```

Comment:
were already at a stage where the reset was called 
so we dont really have to check if a task should be restarted at this point 

i would recommend querying for all failed tasks on line 2542 if resetwhenfailed is true and leaving it be otherwise and appending all of them at once

Reply:
Does the db query for failed exec tasks check for the same status as the evergreen.IsfailedTaskStatus call (going to follow up on this thread once I research)? If they're different, I wonder which set of exec tasks we want

Reply:
Agreed that we could just be querying for the wanted exec tasks based on the display task fields (in which case maybe this helper function isn't needed)

Reply:
I looked in to the queries and it does take in to account the different failure status like my helper-- migrating to do a conditional to correctly query all or only exec!